Zend/zend_vm_def.h

```diff
@@ -9544,6 +9544,136 @@ ZEND_VM_HANDLER(202, ZEND_CALLABLE_CONVERT, UNUSED, UNUSED)
 	ZEND_VM_NEXT_OPCODE();
 }
 
+ZEND_VM_HANDLER(204, ZEND_FRAMELESS_ICALL_0, UNUSED, UNUSED)
+{
+	USE_OPLINE
+	SAVE_OPLINE();
+	zend_frameless_function_0 function = (zend_frameless_function_0)ZEND_FLF_HANDLER(opline);
+	ZVAL_NULL(EX_VAR(opline->result.var));
+	function(EX_VAR(opline->result.var));
+	if (EG(exception)) {
+		HANDLE_EXCEPTION();
+	}
+	ZEND_VM_NEXT_OPCODE();
+}
+
+ZEND_VM_HANDLER(205, ZEND_FRAMELESS_ICALL_1, CONST|TMP|VAR|CV, UNUSED)
+{
+	USE_OPLINE
+	SAVE_OPLINE();
+	zend_frameless_function_1 function = (zend_frameless_function_1)ZEND_FLF_HANDLER(opline);
+	ZVAL_NULL(EX_VAR(opline->result.var));
+
+	zval *arg1;
+	if (OP1_TYPE == IS_CV) {
+		/* Inlined GET_OPx_ZVAL_PTR_DEREF to nest the EG(exception) check. */
+		arg1 = EX_VAR(opline->op1.var);
+		if (UNEXPECTED(Z_TYPE_P(arg1) == IS_UNDEF)) {
+			arg1 = zval_undefined_cv(opline->op1.var EXECUTE_DATA_CC);
+			if (EG(exception)) {
+				FREE_OP1();
+				HANDLE_EXCEPTION();
+			}
+		}
+		ZVAL_DEREF(arg1);
+	} else {
+		arg1 = GET_OP1_ZVAL_PTR_DEREF(BP_VAR_R);
+	}
+
+	function(EX_VAR(opline->result.var), arg1);
+	FREE_OP1();
+	if (EG(exception)) {
+		HANDLE_EXCEPTION();
+	}
+	ZEND_VM_NEXT_OPCODE();
+}
+
+ZEND_VM_HANDLER(206, ZEND_FRAMELESS_ICALL_2, CONST|TMP|VAR|CV, CONST|TMP|VAR|CV)
+{
+	USE_OPLINE
+	SAVE_OPLINE();
+	zend_frameless_function_2 function = (zend_frameless_function_2)ZEND_FLF_HANDLER(opline);
+	ZVAL_NULL(EX_VAR(opline->result.var));
+
+	zval *arg1;
+	if (OP1_TYPE == IS_CV) {
+		/* Inlined GET_OPx_ZVAL_PTR_DEREF to nest the EG(exception) check. */
+		arg1 = EX_VAR(opline->op1.var);
+		if (UNEXPECTED(Z_TYPE_P(arg1) == IS_UNDEF)) {
+			arg1 = zval_undefined_cv(opline->op1.var EXECUTE_DATA_CC);
+			if (EG(exception)) {
+				FREE_OP1();
+				FREE_OP2();
+				HANDLE_EXCEPTION();
+			}
+		}
+		ZVAL_DEREF(arg1);
+	} else {
+		arg1 = GET_OP1_ZVAL_PTR_DEREF(BP_VAR_R);
+	}
+
+	zval *arg2;
+	if (OP2_TYPE == IS_CV) {
+		/* Inlined GET_OPx_ZVAL_PTR_DEREF to nest the EG(exception) check. */
+		arg2 = EX_VAR(opline->op2.var);
+		if (UNEXPECTED(Z_TYPE_P(arg2) == IS_UNDEF)) {
+			arg2 = zval_undefined_cv(opline->op2.var EXECUTE_DATA_CC);
+			if (EG(exception)) {
+				FREE_OP1();
+				FREE_OP2();
+				HANDLE_EXCEPTION();
+			}
+		}
+		ZVAL_DEREF(arg2);
+	} else {
+		arg2 = GET_OP2_ZVAL_PTR_DEREF(BP_VAR_R);
+	}
+
+	function(EX_VAR(opline->result.var), arg1, arg2);
+	FREE_OP1();
+	/* Set OP1 to UNDEF in case FREE_OP2() throws. */
+	if ((OP1_TYPE & (IS_VAR|IS_TMP_VAR)) != 0 && (OP2_TYPE & (IS_VAR|IS_TMP_VAR)) != 0) {
+		ZVAL_UNDEF(EX_VAR(opline->op1.var));
+	}
+	FREE_OP2();
+	if (EG(exception)) {
+		HANDLE_EXCEPTION();
+	}
+	ZEND_VM_NEXT_OPCODE();
+}
+
+ZEND_VM_HANDLER(207, ZEND_FRAMELESS_ICALL_3, ANY, ANY)
```

Comment:
What is the reason to avoid specialization?

Reply:
To avoid bloating the VM, but I did not measure how big the difference is. Which operands would you specialize? Just op1/2 or also op_data? `CONST|TMP|VAR|CV` like the others? That sounds like a lot.

Reply:
OK. Specialization itself is not a problem. The real problem that undefined IS_CV may be handled improperly.